code/globalincs/pstypes.h

```diff
@@ -20,6 +20,8 @@
 #include "globalincs/vmallocator.h"
 #include "utils/strings.h"
 
+#include "imgui.h"
```

Comment:
Is there a pressing need to have an imgui include this far up in the chain, instead of, say, in the cpp-files of the lab itself?
99% of the code will not need it, but it will basically pollute the namespace with imgui stuff for (almost) every file.

Reply:
Yes, unfortunately. Dear Imgui clashes with our custom allocator, requiring a special case to be added later in that same file.

Reply:
If that is just for the declaration of ``ImDrawListSplitter``, couldn't you just forward declare the class? Since it's just used as a pointer, an incomplete type should suffice here

Reply:
Problem is that that doesn't work, at least on MSVC. Trying to forward declare this struct just leads to compiler errors.

Reply:
I just tested this. There is a way to solve this properly (tested on a local checkout):
Add an enable_if to the template memset method we have and use the forward declare. This prevents the ImGui stuff from seeing this function overload which was for some reason preferred to the more specified function overload for the normal forward declare:
```cpp
template<typename T>
void *memset_if_trivial_else_error(typename std::enable_if<!std::is_same<T, ImDrawListSplitter>::value, T>::type *memset_data, int ch, size_t count)
{
```

Reply:
still fails on my end: ```C:\Users\fwolt\Source\fs2open.github.com\code\globalincs\pstypes.h(497): error C2338: static_assert failed: 'memset on non-trivial object'
  C:\Users\fwolt\Source\fs2open.github.com\lib\imgui\imgui.h(2456): note: see reference to function template instantiation 'void *std::memset_if_trivial_else_error<ImDrawListSplitter>(T *,int,size_t)' being compiled
          with
          [
              T=ImDrawListSplitter
          ]```

Reply:
Just to make sure, you replaced the already templated memset and not the ImGui memset, right?
On MSVC2017, the following compiles when replacing the memset section, and with the imgui include in pstypes.h removed
```cpp
struct ImDrawListSplitter;

// Put into std to be compatible with code that uses std::mem*
namespace std
{
	template<typename T>
	using trivial_check = std::is_trivially_copyable<T>;

	template<typename T>
	void* memset_if_trivial_else_error(typename std::enable_if<!std::is_same<T, ImDrawListSplitter>::value, T>::type* memset_data, int ch, size_t count)
	{
		static_assert(trivial_check<T>::value, "memset on non-trivial object");
		return ptr_memset(memset_data, ch, count);
	}

	// assume memset on a void* is "safe"
	// only used in cutscene/mveplayer.cpp:mve_video_createbuf()
	inline void *memset_if_trivial_else_error(void *memset_data, int ch, size_t count)
	{
		return ptr_memset(memset_data, ch, count);
	}

	// Dear ImGui triggers these as well, so we need to let them pass
	inline void *memset_if_trivial_else_error(ImDrawListSplitter *memset_data, int ch, size_t count)
	{
		return ptr_memset(memset_data, ch, count);
	}
```
I still get some (probably unrelated) warnings from the lab code, but nothing that prevents compilation outright

Reply:
ahh, that's what I was missing. Works now.